src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -940,20 +970,49 @@ where
         // The global view of the source_upper, which we track by combining
         // summaries from the raw reader operators.
         let mut global_source_upper = OffsetAntichain::new();
+        // For each partition, track the batch time and offset of the most recent
+        // batch. When the `batch_input` frontier has advanced past this batch time
+        // the `global_source_upper` can be advanced.
+        let mut batch_times: HashMap<PartitionId, (Timestamp, MzOffset)> = HashMap::new();
+
         let mut untimestamped_batches = VecDeque::new();
 
         move |frontiers| {
-            batch_input.for_each(|_cap, data| {
+            batch_input.for_each(|cap, data| {
                 data.swap(&mut batch_buffer);
                 for batch in batch_buffer.drain(..) {
                     let batch = batch
                         .borrow_mut()
                         .take()
                         .expect("batch already taken, but we should be the only consumer");
                     for (pid, offset) in batch.source_upper.iter() {
-                        let previous_offset = global_source_upper.insert(pid.clone(), *offset);
-                        if let Some(previous_offset) = previous_offset {
-                            assert!(previous_offset <= *offset);
+                        // If we have a time for the batch in `batch_times` already, we go onto
+                        // checking the frontier. Otherwise, place the time and offset in the map
+                        // and wait for the next empty batch to push us along.
+                        if let Some((batch_ts, advance_to)) = batch_times.get(pid) {
+                            trace!(
+                                "reclock({id}) {worker_id}/{worker_count}: batch frontier: {:?}, \
+                                processing batch at ts: {}, \
+                                wanting to advance to: {:?}",
+                                frontiers[0].frontier(),
+                                batch_ts,
+                                advance_to
+                            );
+                            // If the batch_input frontier has advanced past this batch time, then
+                            // we can advance the `global_source_upper`, which will later downgrade
+                            // `cap_set`, as we are sure we have seen all messages for this
+                            // `source_upper`.
+                            if let Some(frontier_ts) = frontiers[0].frontier().as_option() {
```

Comment:
subtlety question 2: do we also need to delay dropping the capability, until the reclock follower can actually reclock these messages? otherwise, we may downgrade the capability before the `ReclockFollower::reclock` actually finishes with the messages? Or is it true that the `global_source_upper` always lags the `remap_frontier`? My guess is that it ISNT true, which means below, we should avoid downgrading until the result of `reclock_frontier` is unambiguously past the `global_source_upper`?